### PR TITLE
docs: add AGENTS.md and CLAUDE.md with agent guidelines under .claude/

### PR DESCRIPTION
**Pull request by ShimonSte**
Introduce agent-facing documentation at the repository root so AI coding agents (and contributors) have a concise, codebase-grounded entry point.

- AGENTS.md: one-page guide covering supported versions, build/test commands, and the hard boundaries that protect cross-version duplication, the v2 Java client contract, public-API names, Jackson shading, VariantType scoping, AutoCloseable handling, push-down semantics, and the CHANGELOG entry rule.
- CLAUDE.md: thin pointer to AGENTS.md plus links into .claude/.
- .claude/code-layout.md: module map, push-down surface as a reference matrix (with explicit non-goals: no projection pushdown beyond column pruning, no join pushdown, no source-side expression evaluation beyond SQLHelper), and the narrow cross-version drift that actually matters when diffing trees (AnalysisException constructor in 4.0, functionRegistry threading in ExprUtils starting at 3.4).
- .claude/build-and-test.md: Cloud SharedMergeTree consistency, Gradle daemon caching when switching profiles, CI matrix detail.
- .claude/hard-rules.md: rationale for every hard rule and the full per-PR checklist.

---

> [!NOTE]
> **Low Risk**
> Documentation-only plus removal of unused imports; no connector behavior or public API changes.
> 
> **Overview**
> Adds **agent-facing documentation** so AI tools and contributors have a single entry point: **`AGENTS.md`** (connector overview, supported matrix, build/test commands, Jackson shading, links to deeper docs), **`CLAUDE.md`** (pointer to `AGENTS.md`), and **`.claude/`** deep dives on **build/test pitfalls** (`-D` vs `-P`, single-suite runs, Cloud consistency, CI), **code layout / push-down / V2 client rules**, and **hard rules plus a per-PR checklist**.
> 
> Separately, removes the unused **V1** `com.clickhouse.client.ClickHouseResponse` import from **`ClickHouseReader.scala`** in all four Spark trees (`3.3`–`4.0`), aligning with the documented V2-only client policy; **no runtime behavior change**.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 32fcd9c90953711b26af35c71d09fe714536f3cf. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: CLAUDE.md [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+-->
+
+# CLAUDE.md
+
+Agent guidelines for this repository live in [`AGENTS.md`](AGENTS.md).
```
Comment (alex-clickhouse) [inline CLAUDE.md]: Reference the file using `@AGENTS.md` instead, it'll include the contents automatically instead of taking an extra turn to read the referenced file.

**Comment by alex-clickhouse** [comment]
In terms of the file layout, there's quite a bit of duplication between AGENTS and the 3 sub-files in .claude/. I think it would be better to deduplicate and put everything in a single file. It's not going to eat much of the context, and it makes sense to have all of that info always available.

If you want to keep the split between files instead of dumping everything in a single one, you could deduplicate the info and then just include the contents of the other files in AGENTS.md by @-mentioning them.

File: .claude/hard-rules.md [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+      (`Added` / `Changed` / `Fixed` / `Removed`) with a PR link.
+- [ ] Breaking changes (default flips, removed configs, renamed classes,
+      changed type mappings) are called out explicitly in the PR
+      description **and** in `CHANGELOG.md` under a **Breaking Changes**
+      sub-section.
+- [ ] Tests cover the happy path **and** the non-trivial edge cases the
+      change actually exposes (nulls, empty partition, type boundary,
+      distributed write, etc.). New push-down code has a test that verifies
+      the generated SQL.
+- [ ] Commit message follows [Conventional Commits](https://www.conventionalcommits.org/)
```
Comment (alex-clickhouse) [inline .claude/hard-rules.md:119]: Two things that might be worth adding to this workflow checklist:
1. Request a review from a subagent before committing, then evaluate and react to the feedback.
2. If you have an easy way to check test coverage locally, ask it to run it and look at the results to make sure coverage is solid.

File: AGENTS.md [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+limitations under the License.
+-->
+
+# Agent Guidelines — Spark ClickHouse Connector
+
+## What this is
+
+- Spark DataSource V2 connector for ClickHouse.
+- Talks to ClickHouse over HTTP via `com.clickhouse:client-v2`. **No JDBC, no
+  V1 client.**
```
Comment (BentsiLeviav) [inline AGENTS.md]: `No JDBC` - it is not accurate. Every version's `build.gradle` has `testImplementation("com.clickhouse:clickhouse-jdbc:...")` and each explicitly excludes it as a transitive runtime dep: `exclude group: "com.clickhouse", module: "clickhouse-jdbc"`

```suggestion
- Talks to ClickHouse over HTTP via `com.clickhouse:client-v2`. **No JDBC at runtime (JDBC appears only as a test dependency), no V1 client.**
```

File: .claude/build-and-test.md [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+default (4.0 / 2.13) instead of 3.5, with no error. Always pass
+`-Dspark_binary_version=...` and `-Dscala_binary_version=...` explicitly.
+
+**Subproject names embed the Scala suffix:** `:clickhouse-spark-<v>_<scala>:`,
+`:clickhouse-spark-it-<v>_<scala>:`, `:clickhouse-spark-runtime-<v>_<scala>:`
+— not `:spark-<v>:clickhouse-spark:`. Run `./gradlew projects` to see the live
+list. Direct task example:
+
+```bash
+./gradlew :clickhouse-spark-3.5_2.13:test --tests=ClickHouseGenericSuite \
```
Comment (BentsiLeviav) [inline .claude/build-and-test.md]: * Are these test parameters intentional? And this specific project and test? When I run this, I get `No tests were executed.`

* Why not use assemble?

File: .claude/build-and-test.md [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+
+## Tests
+
+Tag-partitioned:
+
+- Default `./gradlew test` excludes `org.scalatest.tags.Slow` and
+  `org.scalatest.tags.Cloud`.
+- `./gradlew slowTest` — includes Slow-tagged tests (e.g. `TPCDSClusterSuite`).
+- `./gradlew cloudTest` — runs against a real ClickHouse Cloud instance.
+  Requires `CLICKHOUSE_CLOUD_HOST` / `_USER` / `_PASSWORD` (optionally
```
Comment (BentsiLeviav) [inline .claude/build-and-test.md]: ` _USER` defaults to "default" - it belongs in the optional group with the ports, not the required list. https://github.com/ClickHouse/spark-clickhouse-connector/blob/f54ed8dcd410cf1a202656dd1894889c6631b5da/clickhouse-core/src/testFixtures/scala/com/clickhouse/spark/base/ClickHouseCloudMixIn.scala#L28

File: .claude/build-and-test.md [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+call `./gradlew --stop` between profiles to evict the daemon. `--rerun-tasks`
+alone does **not** address this.
+
+Switching from Scala 2.13 to 2.12 (i.e. building 3.3 after 3.5) can also
+leave stale 2.13 `.class` files in `clickhouse-core/build/` that the 2.12
+compiler cannot read. On an unexplained Scala 2.12 compile failure in
+`clickhouse-core` after a profile switch:
+
+```bash
+rm -rf clickhouse-core/build/classes clickhouse-core/build/libs
```
Comment (BentsiLeviav) [inline .claude/build-and-test.md]: Can you share why not to use `./gradlew clean`?

File: .claude/build-and-test.md [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+   `insert_quorum_parallel=0`** on SharedMergeTree to take effect.
+
+`Thread.sleep(...)` is not an acceptable primary strategy. Existing Cloud
+tests in `SparkClickHouseSingleTest` and related fixtures still use bare
+`Thread.sleep(2000)` guarded by `if (isCloud) ...` — that pattern is the
+legacy approach this rule is intended to phase out, not a model to copy.
+
+### Profile-switching pitfalls in one shell
+
+The Gradle daemon caches `System.getProperty` values from the JVM it started
```
Comment (BentsiLeviav) [inline .claude/build-and-test.md]: Is that supposed to be consistent/reproducible?

Couldn't reproduce it with Claude

File: .claude/build-and-test.md [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+  break the 2.12 lane.** Watch for: string-interpolator extractors on the
+  LHS of `case` (`case s"prefix-$x" =>`), the `LazyList` rename, certain
+  collection methods (`.iterableFactory`, etc.). Neither Spotless nor the
+  IDE will flag these; only the 2.12 CI matrix entries will.
+
+---
+
+## CI quick reference
+
+Workflows in `.github/workflows/`:
```
Comment (BentsiLeviav) [inline .claude/build-and-test.md:188]: Should we list all workflows? If not - why?

File: .claude/code-layout.md [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+| Function lookup in expressions | `ExprUtils.toCatalyst` / `toSparkExpression` / `toSparkTransformOpt` + `FunctionRegistry` | V2 `ApplyTransform` whose name resolves in the `FunctionRegistry` is recognised; args walk recursively. |
+| DDL `PARTITION BY` rendering | `ExprUtils.toClickHouse(transform: Transform): Expr` | Used at table create/alter time. Identifier quoting is the user's responsibility (DDL is user-typed). Do not "fix" this in lockstep with read-path quoting. |
+| Catalog functions (`SHOW FUNCTIONS`, etc.) | `ClickHouseCatalog.{listFunctions, loadFunction}` (`extends FunctionCatalog`) | Backed by `StaticFunctionRegistry` (built-in hash funcs) + a `DynamicFunctionRegistry` populated at `initialize` (per-cluster shard hash funcs, e.g. `clickhouse_shard_xxHash64`). |
+
+**Explicit non-goals (and how Spark falls back):**
+
+- **No projection push-down.** Spark V2 has no `SupportsPushDownProjection`
```
Comment (BentsiLeviav) [inline .claude/code-layout.md:97]: Since [this](https://github.com/ClickHouse/spark-clickhouse-connector/pull/543) was merged, should we add `SupportsPushDownTopN`?

File: .claude/code-layout.md [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+- **`ExprUtils` function-translation signatures differ between 3.3 and
+  3.4+.** 3.4 / 3.5 / 4.0 thread a `functionRegistry: FunctionRegistry`
+  argument through `toCatalyst`, `toSparkTransformOpt`, `toSparkExpression`,
+  and `toClickHouse`. 3.3 has no `FunctionRegistry` parameter on these
+  signatures — `ApplyTransform` resolution is narrower as a result, and any
+  push-down that wants to translate a Spark function call into a ClickHouse
+  one must implement the 3.3 path more conservatively. Do not paste a 3.4+
+  patch into 3.3 without auditing every call site.
+- **`SupportsPushDown*` mix-in list on `ClickHouseScanBuilder` is uniform
+  on `main` today** (Limit + Filters + Aggregates + RequiredColumns) but
```
Comment (BentsiLeviav) [inline .claude/code-layout.md]: Should we include TopN here as well?

File: .claude/hard-rules.md [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+repeat.
+
+1. **Don't edit only one Spark version.** If a file exists in multiple
+   `spark-<v>/` trees, decide upfront which versions you'll touch. The CI
+   matrix catches mismatches and the human reviewer is annoyed by them. See
+   [`code-layout.md`](code-layout.md#cross-version-duplication-is-intentional).
+
+2. **Don't use Spark API symbols that don't exist in all targeted versions.**
+   Check the Scaladoc for the lowest Spark version you're modifying.
+   See [`code-layout.md`](code-layout.md#push-down-api-drift).
```
Comment (BentsiLeviav) [inline .claude/hard-rules.md]: `push-down-api-drift` doesn't exist

**Review by BentsiLeviav — CHANGES_REQUESTED** [review]

File: .claude/build-and-test.md (hunk 4 above)
Comment (ShimonSte) [reply .claude/build-and-test.md]: The point of this bash was to show the agent how to address a single subproject by its full _<scala> name.

Anyway i changed it around a bit and added the test to the testing section.

File: .claude/build-and-test.md (hunk 8 above)
Comment (ShimonSte) [reply .claude/build-and-test.md:188]: I wanted to add only ones that potentially anyone can use (no publishing workflows etc.)
Added the rest as i want the agents md to have the full info also for us

**Comment by ShimonSte** [comment]
bugbot run verbose=true

**Review by BentsiLeviav — APPROVED** [review]
